processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java

```diff
@@ -49,16 +51,41 @@
  */
 public class MapperConfiguration {
 
+    private static Map<Element, MapperConfiguration> configurations = new WeakHashMap<Element, MapperConfiguration>();
+
     private final MapperPrism mapperPrism;
+    private final AnnotationMirror componentModelAnnotation;
     private final MapperConfigPrism mapperConfigPrism;
     private final DeclaredType config;
 
-    public static MapperConfiguration getInstanceOn(Element e) {
-        return new MapperConfiguration( MapperPrism.getInstanceOn( e ) );
+    private final FormattingMessager messager;
+
+    public static MapperConfiguration getInstanceOn(Element e, FormattingMessager messager) {
+        // store configurations per Element instance in a cache map
+        // this prevents creating many instances of this configuration for the same Element
+        // it will be assumed that this method will never be called with different messagers for the same Element
+        // one reason to cache instances is that the the MapperConfiguration validates the Element and multiple
+        // instantiations would lead to multiple identical warnings/errors
```

Comment:
Is this okay? Maybe a `WeakHashMap` is too much as a processor is only a short-running process. Maybe you've got better ideas